### PR TITLE
ArcGIS Dataset picker redesign

### LINKED ISSUE #146: Create Search / filtering panel for ESRI integration
As we add datasets to the ESRI integration, we need to make sure we provide the search & filtering functionality to let users better decide what they want to see on the map.

### PR DESCRIPTION
**Pull request by bhousel**
closes #146 

Now that we have a wealth of datasets to choose from in the ArcGIS portal, the view-manage dataset screen needs a redesign.

## Existing screen

We use dots to represent "pages" of data, and each page shows 4 datasets.  We've outgrown this design:

Screen Shot 2021-05-06 at 11 17 16 AM

## New Screen (This PR)

The dots are gone and the user can now type whatever they are looking for in a "filter datasets" input.  The user can also filter by category (addresses, buildings, featured, preview). These categories are just whatever we get from ArcGIS, so we can always add more or recategorize as needed.

Also added a "Clear Filters" link and "# dataset(s) found" to make it a bit more easy to use.

datasets

I snuck one other feature in after talking to our testers:
- [x] When adding any of the Esri "buildings" datasets, we uncheck the magenta Microsoft Buildings layer to avoid map clutter.  (The user can turn it back on if they really want to)

A few other followup things that I might do before merging or as followup work later:
- It would be nice to show "Datasets near the current map position" or some kind of high level map of where the data is.  A map with bounding boxes would be cool.  
- [x] The unstyled scrollbar is bugging me so I'll probably just style it to be like the other dark RapiD UI components.
- [x] We should cut off the datasets at some threshold.  We have 63 datasets now, and this would work fine up into the 100s, but not sure how it would feel past that.  So I'll probably clamp the results at 100, and write some code to say ">100 dataset(s) found".
- [x] We have a "featured" category that we're not using.  Maybe we should stick featured datasets at the top, or display a ⭐ or something.

### CODE REVIEW

**Comment by bhousel** [comment]
I'm pretty happy with how it ended up now - @bonkles want to take a review pass?

**Comment by Bonkles** [comment]
One smol bit of text indentation / formatting feedback- we probably want to break up the star emoji and the text label for 'Featured'. In certain contexts the star can overrun the text kerning and look a little bit smooshed. Is this a nit? You bet your SWEET CAKES it is.  

If we want to fix this in a very low-tech way, we can add `letter-spacing: .5px` to the .rapid-view-manage-dataset-featured class, which takes us from this: 
image

To this: 
image

But I suspect it's better to simply encase the emoji in its own bit of markup and style that as we see fit.

**Comment by bhousel** [comment]
> But I suspect it's better to simply encase the emoji in its own bit of markup and style that as we see fit.

Very good point!  Yeah the ⭐ Featured badge was totally an afterthought and it looks that way 😆 
I'll add a span and give it a bit of extra space.

File: modules/ui/rapid_view_manage_datasets.js [hunk 1]
```diff
@@ -154,17 +123,70 @@ export function uiRapidViewManageDatasets(context, parentModal) {
...
+        _filterText = val.trim().toLowerCase();
+        dsSection.call(renderDatasets);
+      });
+
+
+    let filterTypeEnter = filterEnter
+      .append('div')
+      .attr('class', 'rapid-view-manage-filter-type-wrap');
+
+    filterTypeEnter
```
Comment (Bonkles) [inline modules/ui/rapid_view_manage_datasets.js:158]: I noticed that the filterType combo box lets users enter whatever string they want, but I am not clear on whether we want that. It seems like we should just pre-fill the box with the known types that we infer from the datasets, and not let the user free-edit the text in it. 

Here's a screenie I took of a user typing in 'roads' thinking that they are, in fact, quite clever. The form does an excellent job letting them know they asked for something that doesn't exist, but I'm wondering if we even need to  give them that freedom. 

image

File: modules/ui/rapid_view_manage_datasets.js [hunk 2]
```diff
@@ -216,30 +228,88 @@ export function uiRapidViewManageDatasets(context, parentModal) {
...
         .then(results => {
-          // exclude preview datasets unless user has opted into them
-          return _datasetInfo = Object.values(results)
-            .filter(d => showPreview || !d.groupCategories.some(category => category === '/Categories/Preview'));
+          // Build set of available categories
+          let categories = new Set();
+
+          Object.values(results).forEach(d => {
+            d.groupCategories.forEach(c => {
+              categories.add(c.toLowerCase().replace('/categories/', ''));
```
Comment (Bonkles) [inline modules/ui/rapid_view_manage_datasets.js:263]: Ah, this is swank. I see you're constructing the categories on-the-fly already, this is excellent!

File: modules/ui/rapid_view_manage_datasets.js [hunk 3]
```diff
@@ -216,30 +228,88 @@ export function uiRapidViewManageDatasets(context, parentModal) {
...
 
-    selection.text('');
+    results.classed('hide', false);
+    status.classed('hide', true);
+
+    // Apply filters
+    let count = 0;
+    _datasetInfo.forEach(d => {
+      count++;
+      const title = (d.title || '').toLowerCase();
```
Comment (Bonkles) [inline modules/ui/rapid_view_manage_datasets.js:293]: I see you have done case-insensitive compares... I see this is NOT your first text filtering rodeo. Well played. I will get you next time...

File: modules/ui/rapid_view_manage_datasets.js [hunk 4]
```diff
@@ -361,10 +447,15 @@ export function uiRapidViewManageDatasets(context, parentModal) {
         dataset.extent = geoExtent(d.extent);
       }
 
-      // Test running building layers only through conflation service
+      // Test running building layers through FBML conflation service
       if (isBuildings) {
         dataset.conflated = true;
         dataset.service = 'fbml';
+
+        // and disable the Microsoft buildings to avoid clutter
```
Comment (Bonkles) [inline modules/ui/rapid_view_manage_datasets.js:455]: I loooove this addition.

File: modules/ui/rapid_view_manage_datasets.js [hunk 5]
```diff
@@ -383,5 +474,18 @@ export function uiRapidViewManageDatasets(context, parentModal) {
   }
 
 
+  function highlight(needle, haystack) {
```
Comment (Bonkles) [inline modules/ui/rapid_view_manage_datasets.js:477]: best variable names evar?

File: modules/ui/rapid_view_manage_datasets.js [hunk 6]
```diff
@@ -154,17 +123,70 @@ export function uiRapidViewManageDatasets(context, parentModal) {
...
+      });
+
+
+    let filterTypeEnter = filterEnter
+      .append('div')
+      .attr('class', 'rapid-view-manage-filter-type-wrap');
+
+    filterTypeEnter
+      .append('input')
+      .attr('class', 'rapid-view-manage-filter-type')
```
Comment (Bonkles) [inline modules/ui/rapid_view_manage_datasets.js:160]: Building on this- if the user enters some text in the 'type' field, hits enter, sees '0 dataset(s) found', and then types some more into the text field, we should give them a visual cue that they haven't yet hit 'entier' yet. perhaps we grey out the '0 datasets found' and / or replace it with an entreaty to 'hit enter'? dunno. Moot point if we just want to eliminate the free-form text edit nature of the control.

**Review by Bonkles — CHANGES_REQUESTED** [review]

**Comment by Bonkles** [comment]
Would I recommend this changeset to other users of RapiD? I would. No issues with code or logical structure, just the UI type things I found above.

File: modules/ui/rapid_view_manage_datasets.js (hunk 1 above)
Comment (bhousel) [reply modules/ui/rapid_view_manage_datasets.js:158]: > It seems like we should just pre-fill the box with the known types that we infer from the datasets, and not let the user free-edit the text in it.

I agree - I think I can switch the combo to work that way..

**Comment by bhousel** [comment]
Thanks for the feedback @Bonkles!  I'll take care of these and then merge..

- [ ] cleanup the <sub><sub>feature</sub></sub> badge
- [ ] don't allow free text entry in filterType combo

**Comment by bhousel** [comment]
I made the titles slightly bigger, added some space around the badge and some internal margin around the star and label:

Screen Shot 2021-05-13 at 10 15 10 AM

It's a lot of information to display but it feels a bit less claustrophobic.. 😅

**Comment by bhousel** [comment]
Hey @Bonkles - I think it's ready to merge now!

**Review by Bonkles — APPROVED** [review]
This. Looks. Dope.

**Pull request merged**